### PR TITLE
Python 3 compatibility

### PR DESCRIPTION
**Pull request by takluyver**
This gets PyMC running on Python 3. The diff is huge, but very little of it is real structural changes.

There's now just a couple of test failures left, both on Python 2 and Python 3, and they were there before I started.

It's using [six](http://packages.python.org/six/), to give one codebase that installs on recent versions of Python 2 and 3. This time, I'm not having 2to3 run at setup, so the Python code needs to be valid Python 2 and 3.

I've also found and fixed a couple of string exceptions (`raise "message"`), which are invalid on recent Python 2 versions as well.

### CODE REVIEW

**Comment by takluyver** [comment]
OK, we're getting there. The test suite on Python 2.7 is now back to the state it was in before I started, and there's just a handful more errors on Python 3.2.

**Comment by takluyver** [comment]
I think this is ready for review. There are now fewer tests failing on Python 3 than on Python 2 (3 or 4 tests were failing before I did anything).

File: pymc/gp/step_methods.py [hunk 1]
```diff
@@ -81,15 +88,13 @@ def competence(stochastic, metro_class=metro_class):
...
-    wrapper.__doc__ = """A modified version of class %s that handles parents of Gaussian processes.
-Docstring of class %s: \n\n%s"""%(metro_class.__name__,metro_class.__name__,metro_class.__doc__)
             
     return wrapper
 
 
 # Wrap all registered Metropolis step methods to use GP parents.
 new_sm_dict = {}
-filtered_registry = filter(lambda x: issubclass(x, pm.Metropolis), pm.StepMethodRegistry)
+filtered_registry = [sm for sm in pm.StepMethodRegistry if issubclass(sm, pm.Metropolis)]
```
Comment (takluyver) [inline pymc/gp/step_methods.py:97]: Out of interest: this seemingly innocuous change was the trickiest bit of porting to Python 3. StepMethods are automatically added to StepMethodRegistry when they're created, by a metaclass. This code is selecting some items from the registry to subclass. But in Python 3, `filter()` returns an iterator, not a list. The new subclasses got added to the registry before the filter had finished, so they too got subclassed, and so on. Memory usage very quickly exploded, my computer started swapping, and I had to do a hard reset.

Wrapping the `filter()` call in a `list()` would have sufficed, but I thought the list comprehension was clearer.

**Comment by fonnesbeck** [comment]
Thanks for working on this. I'm just back from holiday, so give me a day or two to check this out.

**Comment by takluyver** [comment]
@fonnesbeck : Just checking in on this - there's no rush, but obviously the longer it waits, the more merge conflicts will arise.

**Comment by fonnesbeck** [comment]
OK, I was able to merge this manually. Looks good -- the unit tests run clean.

**Comment by takluyver** [comment]
Great, thanks Chris.

**Pull request merged**